LuaVAO Instanced Models

Krogoth100 opened the pull request:
+vao:ModifyUnitBins
+vao:ModifyFeatureBins
+vao:SubmitBins

Bins are placed in a vector instead of using map directly, it should have some advantages (For one, vector is transparent)

As of now it is simplified: only InstanceData, no other instance attributes. I will be expanding it to work with arbitrary instance data a bit later.

W.I.P.

Beherith commented:
How does this relate to the push-pop of instanceVBOidTable ?

lhog commented:
A few comments:

- I think it's better to stick to separate Add/Remove functions
- Why did you add explicit (but optional) count of elements?

lhog commented on rts/Lua/LuaVAOImpl.cpp:
```diff
@@ -185,6 +194,116 @@ int LuaVAOImpl::AddObjectsToSubmissionImpl(const sol::stack_table& ids)
+	const size_t addedObjectCount = (addedCount.has_value()? addedCount.value() : addedObjects.size());
+	if (removedObjectCount == 0 && addedObjectCount == 0) return;
+
+	size_t firstChangedBinIndex = -1;
```
This tripped me up a bit until I saw it was size_t, better go for `std::numeric_limits<size_t>::max()`

Krogoth100 replied:
Ok

lhog commented on rts/Lua/LuaVAOImpl.cpp:
```diff
@@ -185,6 +194,116 @@ int LuaVAOImpl::AddObjectsToSubmissionImpl(const sol::stack_table& ids)
+		bin.instanceData.emplace_back(
+			static_cast<uint32_t>(matIndex),
+			static_cast<uint8_t>(obj->team), // initial team id, will not be updated (unreliable: shader must obtain it somehow else, otherwise it will not reflect transfer to another team)
+			static_cast<uint8_t>(0), // no draw flags, they will not be updated
```
Please still submit initial drawflags in order to adhere to the existing conventions.

lhog commented on rts/Lua/LuaVAOImpl.cpp:
```diff
@@ -185,6 +194,116 @@ int LuaVAOImpl::AddObjectsToSubmissionImpl(const sol::stack_table& ids)
+template<typename TObj>
+void LuaVAOImpl::Bins::ModifyImpl(const sol::stack_table& removedObjects, const sol::stack_table& addedObjects,
+	sol::optional<size_t> removedCount, sol::optional<size_t> addedCount,
+	std::vector<SDrawElementsIndirectCommand>& submitCmds)
```
This should belong to the class and accessed directly, no need to carry it along with arguments.

Krogoth100 replied:
> This should belong to the class and accessed directly, no need to carry it along with arguments.

The thing is, submitCmds is the only thing that is used from LuaVAO here. All the other fields are good to work this way.
Maybe move this thing in another cpp altogether?

lhog commented:
The implementation generally looks bulkier than it should, are you sure you can't reuse bits and pieces from https://github.com/beyond-all-reason/spring/blob/BAR105/rts/Rendering/Models/3DModelVAO.cpp#L270-L304 ?

Krogoth100 commented:
> How does this relate to the push-pop of instanceVBOidTable ?

I think, this serves same purpose, except it is fully instanced. Basic functions are not instanced in a sense that they do not instance index data, they push indices every time a repetitive model is pushed.

Pick objects via RenderXDrawFlagChanged > process > send to bins of respective VAOs.

Krogoth100 commented:
> The implementation generally looks bulkier than it should, are you sure you can't reuse bits and pieces from https://github.com/beyond-all-reason/spring/blob/BAR105/rts/Rendering/Models/3DModelVAO.cpp#L270-L304 and https://github.com/beyond-all-reason/spring/blob/BAR105/rts/Rendering/Models/3DModelVAO.cpp#L344-L389 ?

First yes, I should probably make instance data external function.
Second has significant differences, hard to reuse.

Krogoth100 commented:
> A few comments:
> 
> * I think it's better to stick to separate Add/Remove functions
> * Why did you add explicit (but optional) count of elements?

Thought of that, that doesn't make practical sense. Usual routine is to collect added and removed objects and just "push changes". No separate Add/Remove make much meaning.

Explicit count because usually Lua knows it, it builds the arrays, no need to call size().

Krogoth100 commented:
@lhog
Addressed some of the issues

+GetObjectInstanceData
a useful one, but I don't know where to put it

I am thinking of separating Bins into another file, not sure how yet, because the code is gonna grow.
Also I want ModifyImpl to be a part of Bins and not of general Vao, as it logically is.

Krogoth100 commented:
@lhog
Planned to do more, but I think that's it for this PR
Only need to sort out GetObjectInstanceData

lhog commented:
This is better, but Add/Remove is a must otherwise it creates disconnect to the current API, also remove element counts, it's a potential source of errors, that doesn't bring any tangible value.
